File: bench/cpp/rs_encode.cpp
// Reed-Solomon erasure encoder — exact algorithm port from src/reed_solomon.rs.
// GF(256), primitive polynomial 0x11D.
// Cauchy generator matrix: coeffs[i*d+j] = 1 / (i ^ (p + j)), where the two
// index sets are disjoint so no denominator is zero. This mirrors
// MatrixKind::Cauchy on the Rust side; see that type's docs for why every
// square submatrix of it is invertible, which the earlier alpha^(i*j)
// construction did not guarantee.
// encode_into: parity[i][k] ^= mul(coeffs[i][j], data[j][k])
//
// Build: g++ -O3 -march=native -std=c++17 -o rs_encode bench/cpp/rs_encode.cpp
// Usage: ./rs_encode [bench/results dir]
// Output: bench/results/cpp.json, bench/results/cpp.checksum

#include <cstdint>
#include <cstring>
#include <string>
#include <vector>
#include <array>
#include <chrono>
#include <cstdio>
#include <cassert>
#include <filesystem>

static constexpr int DATA_SHARDS   = 10;
static constexpr int PARITY_SHARDS = 4;
static constexpr uint32_t ITERS    = 500;
static constexpr uint32_t WARMUP   = 50;

// ---- GF(256) tables --------------------------------------------------------

static uint8_t gf_exp[512];
static uint8_t gf_log[256];

static void build_tables() {
    uint8_t x = 1;
    for (int i = 0; i < 255; ++i) {
        gf_exp[i] = x;
        gf_log[x] = (uint8_t)i;
        bool hi = (x & 0x80) != 0;
        x = (uint8_t)(x << 1);
        if (hi) x ^= 0x1D;
    }
    for (int i = 255; i < 512; ++i)
        gf_exp[i] = gf_exp[i - 255];
}

static inline uint8_t gf_mul(uint8_t a, uint8_t b) {
    if (a == 0 || b == 0) return 0;
    return gf_exp[(int)gf_log[a] + (int)gf_log[b]];
}

static inline uint8_t pow_alpha(int power) {
    return gf_exp[((power % 255) + 255) % 255];
}

// Multiplicative inverse in GF(256): alpha^(255 - log a).
static inline uint8_t gf_inv(uint8_t a) {
    return gf_exp[255 - (int)gf_log[a]];
}

// ---- RS encoder ------------------------------------------------------------

struct ReedSolomon {
    int d, p;
    uint8_t coeffs[PARITY_SHARDS * DATA_SHARDS];

    ReedSolomon(int data_shards, int parity_shards)
        : d(data_shards), p(parity_shards) {
        for (int i = 0; i < p; ++i)
            for (int j = 0; j < d; ++j)
                coeffs[i * d + j] = gf_inv((uint8_t)(i ^ (p + j)));
    }

    void encode_into(const uint8_t* const* data, size_t shard_len,
                     uint8_t* parity_out) const {
        // zero parity
        memset(parity_out, 0, (size_t)p * shard_len);

        for (int j = 0; j < d; ++j) {
            const uint8_t* dj = data[j];
            for (int i = 0; i < p; ++i) {
                uint8_t coef = coeffs[i * d + j];
                if (coef == 0) continue;
                uint8_t* row = parity_out + (size_t)i * shard_len;
                if (coef == 1) {
                    for (size_t k = 0; k < shard_len; ++k)
                        row[k] ^= dj[k];
                } else {
                    for (size_t k = 0; k < shard_len; ++k)
                        row[k] ^= gf_mul(coef, dj[k]);
                }
            }
        }
    }
};

// ---- timing ----------------------------------------------------------------

using Clock = std::chrono::steady_clock;

template<typename F>
double time_ns(F&& fn) {
    // warm-up
    for (uint32_t i = 0; i < WARMUP; ++i) fn();
    auto t0 = Clock::now();
    for (uint32_t i = 0; i < ITERS; ++i) fn();
    auto t1 = Clock::now();
    double total_ns = (double)std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();
    return total_ns / ITERS;
}

static double mib_per_s(size_t payload_bytes, double ns_per_iter) {
    return (payload_bytes / ns_per_iter) * 1e9 / (1024.0 * 1024.0);
}

// ---- checksum helper -------------------------------------------------------

static std::string to_hex(const uint8_t* data, size_t len) {
    static const char hex[] = "0123456789abcdef";
    std::string out;
    out.reserve(len * 2);
    for (size_t i = 0; i < len; ++i) {
        out.push_back(hex[data[i] >> 4]);
        out.push_back(hex[data[i] & 0xf]);
    }
    return out;
}

// ---- main ------------------------------------------------------------------

int main(int argc, char** argv) {
    build_tables();

    const char* out_dir = (argc > 1) ? argv[1] : "bench/results";
    std::filesystem::create_directories(out_dir);

    static constexpr size_t SHARD_LENS[] = {256, 1024, 4096, 16384};
    static constexpr int N_LENS = 4;
    static constexpr size_t CHECKSUM_SHARD_LEN = 1024;

    ReedSolomon rs(DATA_SHARDS, PARITY_SHARDS);

    std::string json_records;
    std::string checksum_hex;

    for (int li = 0; li < N_LENS; ++li) {
        size_t shard_len = SHARD_LENS[li];
        size_t payload_bytes = (size_t)DATA_SHARDS * shard_len;

        // Allocate data shards: shard j filled with byte j.
        std::vector<std::vector<uint8_t>> owned(DATA_SHARDS);
        std::vector<const uint8_t*> ptrs(DATA_SHARDS);
        for (int j = 0; j < DATA_SHARDS; ++j) {
            owned[j].assign(shard_len, (uint8_t)j);
            ptrs[j] = owned[j].data();
        }
        std::vector<uint8_t> parity((size_t)PARITY_SHARDS * shard_len);

        double ns = time_ns([&]() {
            rs.encode_into(ptrs.data(), shard_len, parity.data());
        });
        double mib = mib_per_s(payload_bytes, ns);

        char buf[512];
        snprintf(buf, sizeof(buf),
            "  {\"lang\":\"cpp\",\"impl\":\"encode_into\","
            "\"shard_len\":%zu,\"data_shards\":%d,\"parity_shards\":%d,"
            "\"payload_bytes\":%zu,\"ns_per_iter\":%.1f,\"mib_per_s\":%.1f}",
            shard_len, DATA_SHARDS, PARITY_SHARDS, payload_bytes, ns, mib);

        if (!json_records.empty()) json_records += ",\n";
        json_records += buf;

        // Capture checksum at shard_len == CHECKSUM_SHARD_LEN using seed data.
        if (shard_len == CHECKSUM_SHARD_LEN) {
            // seed: shard j has all bytes = (j*3+1) & 0xFF (matches Rust exporter)
            std::vector<std::vector<uint8_t>> seed(DATA_SHARDS);
            std::vector<const uint8_t*> sptrs(DATA_SHARDS);
            for (int j = 0; j < DATA_SHARDS; ++j) {
                uint8_t fill = (uint8_t)(((uint8_t)j * 3u) + 1u);
                seed[j].assign(CHECKSUM_SHARD_LEN, fill);
                sptrs[j] = seed[j].data();
            }
            std::vector<uint8_t> cparity((size_t)PARITY_SHARDS * CHECKSUM_SHARD_LEN);
            rs.encode_into(sptrs.data(), CHECKSUM_SHARD_LEN, cparity.data());
            checksum_hex = to_hex(cparity.data(), cparity.size());
        }
    }

    // Write JSON
    std::string json_path = std::string(out_dir) + "/cpp.json";
    FILE* jf = fopen(json_path.c_str(), "w");
    assert(jf);
    fprintf(jf, "[\n%s\n]\n", json_records.c_str());
    fclose(jf);
    printf("Wrote %s\n", json_path.c_str());

    // Write checksum
    std::string cs_path = std::string(out_dir) + "/cpp.checksum";
    FILE* cf = fopen(cs_path.c_str(), "w");
    assert(cf);
    fprintf(cf, "%s", checksum_hex.c_str());
    fclose(cf);
    printf("Wrote %s\n", cs_path.c_str());

    // Print summary
    printf("\nC++ RS encode throughput (MiB/s):\n");
    printf("%-32s %8s %8s %8s %8s\n", "impl", "256B", "1KiB", "4KiB", "16KiB");
    // Parse from records is awkward; re-run quickly.
    double mibs[N_LENS];
    for (int li = 0; li < N_LENS; ++li) {
        size_t shard_len = SHARD_LENS[li];
        size_t payload_bytes = (size_t)DATA_SHARDS * shard_len;
        std::vector<std::vector<uint8_t>> owned(DATA_SHARDS);
        std::vector<const uint8_t*> ptrs(DATA_SHARDS);
        for (int j = 0; j < DATA_SHARDS; ++j) {
            owned[j].assign(shard_len, (uint8_t)j);
            ptrs[j] = owned[j].data();
        }
        std::vector<uint8_t> parity((size_t)PARITY_SHARDS * shard_len);
        double ns = time_ns([&]() {
            rs.encode_into(ptrs.data(), shard_len, parity.data());
        });
        mibs[li] = mib_per_s(payload_bytes, ns);
    }
    printf("%-32s %8.1f %8.1f %8.1f %8.1f\n",
           "encode_into", mibs[0], mibs[1], mibs[2], mibs[3]);

    return 0;
}

File: bench/python/rs_encode.py
#!/usr/bin/env python3
"""
Cross-language Reed-Solomon benchmark driver.

Produces two result sets:
  1. python_same_algo  — pure-Python port of the EXACT algorithm in
     src/reed_solomon.rs (GF(256) 0x11D, Vandermonde matrix, encode_into).
     Byte-identical to Rust and C++ for the same input; used in the
     correctness-gate checksum comparison.
  2. python_reedsolo   — the `reedsolo` library (generator-polynomial RS,
     different algorithm, different output). Included as an ecosystem
     reference only; its parity bytes will NOT match the same-algo group.

Output: bench/results/python.json
        bench/results/python_same_algo.checksum
"""

import os
import sys
import time
import json

# ---------------------------------------------------------------------------
# GF(256) with primitive polynomial 0x11D
# ---------------------------------------------------------------------------

_EXP = [0] * 512
_LOG = [0] * 256

def _build_tables():
    x = 1
    for i in range(255):
        _EXP[i] = x
        _LOG[x] = i
        hi = (x & 0x80) != 0
        x = (x << 1) & 0xFF
        if hi:
            x ^= 0x1D
    for i in range(255, 512):
        _EXP[i] = _EXP[i - 255]

_build_tables()

def gf_mul(a: int, b: int) -> int:
    if a == 0 or b == 0:
        return 0
    return _EXP[_LOG[a] + _LOG[b]]

def pow_alpha(power: int) -> int:
    return _EXP[power % 255]

def gf_inv(a: int) -> int:
    """Multiplicative inverse in GF(256): alpha^(255 - log a)."""
    return _EXP[255 - _LOG[a]]

# ---------------------------------------------------------------------------
# Same-algorithm encoder (mirrors encode_into from reed_solomon.rs)
# ---------------------------------------------------------------------------

DATA_SHARDS   = 10
PARITY_SHARDS = 4
SHARD_LENS    = [256, 1024, 4096, 16384]
CHECKSUM_SHARD_LEN = 1024

ITERS  = 200   # Python is slow; keep timing runs reasonable
WARMUP = 20

def build_coeffs(d: int, p: int) -> list[int]:
    """Cauchy matrix coeffs[i*d+j] = 1 / (i ^ (p + j)).

    The two index sets are disjoint, so no denominator is zero. This mirrors
    MatrixKind::Cauchy in src/reed_solomon.rs; see that type's docs for why
    every square submatrix of it is invertible, which the earlier alpha^(i*j)
    construction did not guarantee.
    """
    coeffs = []
    for i in range(p):
        for j in range(d):
            coeffs.append(gf_inv(i ^ (p + j)))
    return coeffs

COEFFS = build_coeffs(DATA_SHARDS, PARITY_SHARDS)

def encode_into(data: list[bytearray], parity_out: bytearray, shard_len: int) -> None:
    """Pure-Python port of ReedSolomon::encode_into."""
    d, p = DATA_SHARDS, PARITY_SHARDS
    # zero parity
    for b in range(p * shard_len):
        parity_out[b] = 0

    for j in range(d):
        dj = data[j]
        for i in range(p):
            coef = COEFFS[i * d + j]
            if coef == 0:
                continue
            row_start = i * shard_len
            if coef == 1:
                for k in range(shard_len):
                    parity_out[row_start + k] ^= dj[k]
            else:
                for k in range(shard_len):
                    parity_out[row_start + k] ^= gf_mul(coef, dj[k])

def make_seed_data(shard_len: int) -> list[bytearray]:
    """Deterministic seed: shard j filled with (j*3+1) & 0xFF."""
    return [bytearray([(j * 3 + 1) & 0xFF] * shard_len) for j in range(DATA_SHARDS)]

def time_encode_same_algo(shard_len: int) -> float:
    """Return mean ns/iter for encode_into at this shard_len."""
    data = [bytearray([j] * shard_len) for j in range(DATA_SHARDS)]
    parity = bytearray(PARITY_SHARDS * shard_len)

    for _ in range(WARMUP):
        encode_into(data, parity, shard_len)

    t0 = time.perf_counter_ns()
    for _ in range(ITERS):
        encode_into(data, parity, shard_len)
    t1 = time.perf_counter_ns()
    return (t1 - t0) / ITERS

# ---------------------------------------------------------------------------
# reedsolo baseline
# ---------------------------------------------------------------------------

def time_reedsolo(shard_len: int) -> float | None:
    """Return mean ns/iter for reedsolo encoding of equivalent payload.

    reedsolo works on codewords up to 255 bytes (GF(256)). We use
    RSCodec(nsym=4) — 4 ECC symbols per chunk — and chunk the payload so
    the total number of output parity bytes equals PARITY_SHARDS * shard_len,
    giving a fair comparison of equal-sized output protection.

    Each chunk is at most 251 data bytes (255 - 4 ECC). We tile the payload
    across as many chunks as needed.
    """
    try:
        import reedsolo
    except ImportError:
        return None

    # 4 ECC symbols per chunk; max 251 data bytes per GF(256) codeword.
    nsym = PARITY_SHARDS   # 4
    chunk_data = 255 - nsym  # 251 data bytes per chunk

    payload_bytes = DATA_SHARDS * shard_len
    # pad payload to a multiple of chunk_data
    pad_len = (-payload_bytes) % chunk_data
    data = (bytes(range(256)) * ((payload_bytes + pad_len) // 256 + 1))[:(payload_bytes + pad_len)]

    rs = reedsolo.RSCodec(nsym)

    for _ in range(WARMUP):
        rs.encode(data)

    t0 = time.perf_counter_ns()
    for _ in range(ITERS):
        rs.encode(data)
    t1 = time.perf_counter_ns()
    return (t1 - t0) / ITERS

# ---------------------------------------------------------------------------
# main
# ---------------------------------------------------------------------------

def mib_per_s(payload_bytes: int, ns_per_iter: float) -> float:
    return (payload_bytes / ns_per_iter) * 1e9 / (1024 * 1024)

def main():
    out_dir = "bench/results"
    if len(sys.argv) > 1:
        out_dir = sys.argv[1]
    os.makedirs(out_dir, exist_ok=True)

    records = []
    checksum_hex = None

    print("python_same_algo timings:")
    for shard_len in SHARD_LENS:
        payload_bytes = DATA_SHARDS * shard_len
        ns = time_encode_same_algo(shard_len)
        mib = mib_per_s(payload_bytes, ns)
        print(f"  shard_len={shard_len:6d}  ns/iter={ns:10.0f}  MiB/s={mib:.2f}")
        records.append({
            "lang": "python_same_algo",
            "impl": "encode_into",
            "shard_len": shard_len,
            "data_shards": DATA_SHARDS,
            "parity_shards": PARITY_SHARDS,
            "payload_bytes": payload_bytes,
            "ns_per_iter": round(ns, 1),
            "mib_per_s": round(mib, 1),
        })

        if shard_len == CHECKSUM_SHARD_LEN:
            seed = make_seed_data(CHECKSUM_SHARD_LEN)
            parity = bytearray(PARITY_SHARDS * CHECKSUM_SHARD_LEN)
            encode_into(seed, parity, CHECKSUM_SHARD_LEN)
            checksum_hex = parity.hex()

    # reedsolo
    print("\npython_reedsolo timings:")
    for shard_len in SHARD_LENS:
        payload_bytes = DATA_SHARDS * shard_len
        ns = time_reedsolo(shard_len)
        if ns is None:
            print(f"  shard_len={shard_len:6d}  reedsolo not installed")
            records.append({
                "lang": "python_reedsolo",
                "impl": "reedsolo.RSCodec.encode",
                "shard_len": shard_len,
                "data_shards": DATA_SHARDS,
                "parity_shards": PARITY_SHARDS,
                "payload_bytes": payload_bytes,
                "ns_per_iter": None,
                "mib_per_s": None,
                "note": "reedsolo not installed",
            })
        else:
            mib = mib_per_s(payload_bytes, ns)
            print(f"  shard_len={shard_len:6d}  ns/iter={ns:10.0f}  MiB/s={mib:.2f}")
            records.append({
                "lang": "python_reedsolo",
                "impl": "reedsolo.RSCodec.encode",
                "shard_len": shard_len,
                "data_shards": DATA_SHARDS,
                "parity_shards": PARITY_SHARDS,
                "payload_bytes": payload_bytes,
                "ns_per_iter": round(ns, 1),
                "mib_per_s": round(mib, 1),
                "note": "different algorithm (generator-polynomial); parity bytes do NOT match same_algo group",
            })

    json_path = os.path.join(out_dir, "python.json")
    with open(json_path, "w") as f:
        json.dump(records, f, indent=2)
    print(f"\nWrote {json_path}")

    cs_path = os.path.join(out_dir, "python_same_algo.checksum")
    with open(cs_path, "w") as f:
        f.write(checksum_hex or "")
    print(f"Wrote {cs_path}")

if __name__ == "__main__":
    main()
